pkg/summary/summary.go

```diff
@@ -49,119 +58,144 @@ type Summary struct {
 	Namespaces  []string            `json:"namespaces"`
 }
 
-// Client checks if VPA objects should be created or deleted
-// what? how does it do that? it checks it??
-type Client struct {
-	//changing these two to match the naming in here...but should it be consistent?
-	KubeClient    *kube.ClientInstance
-	KubeClientVPA *kube.VPAClientInstance
-}
+// Summarizer represents a source of generating a summary of VPAs
+type Summarizer struct {
+	Options
 
-var singleton *Client
+	// cached list of vpas
+	vpas []v1beta2.VerticalPodAutoscaler
+}
 
-// GetInstance returns a Client singleton
-func GetInstance() *Client {
-	if singleton == nil {
-		singleton = &Client{
-			KubeClient:    kube.GetInstance(),
-			KubeClientVPA: kube.GetVPAInstance(),
-		}
+// Returns a Summarizer for all goldilocks managed VPAs in all Namespaces
+func NewSummarizer(setters ...Option) *Summarizer {
+	opts := defaultOptions()
+	for _, setter := range setters {
+		setter(opts)
 	}
-	return singleton
-}
 
-// SetInstance sets the singleton using preconstructed k8s and vpa clients. Used for testing.
-func SetInstance(k8s *kube.ClientInstance, vpa *kube.VPAClientInstance) *Client {
-	singleton = &Client{
-		KubeClient:    k8s,
-		KubeClientVPA: vpa,
+	return &Summarizer{
+		Options: *opts,
 	}
-	return singleton
 }
 
-// Run creates a summary of the vpa info for all namespaces.
-func (client *Client) Run(vpaLabels map[string]string, excludeContainers string) (Summary, error) {
-	klog.V(3).Infof("Looking for VPAs with labels: %v", vpaLabels)
-
-	vpaListOptions := metav1.ListOptions{
-		LabelSelector: labels.Set(vpaLabels).String(),
-	}
+// Returns a Summarizer for a known list of VPAs
+func NewSummarizerForVPAs(vpas []v1beta2.VerticalPodAutoscaler, setters ...Option) *Summarizer {
+	summarizer := NewSummarizer(setters...)
 
-	vpas, err := client.KubeClientVPA.Client.AutoscalingV1beta2().VerticalPodAutoscalers("").List(vpaListOptions)
-	if err != nil {
-		klog.Error(err.Error())
-	}
-	klog.V(10).Infof("Found vpas: %v", vpas)
+	// set the cached vpas list directly
+	summarizer.vpas = vpas
 
-	summary, _ := GetInstance().constructSummary(vpas, excludeContainers)
-	return summary, nil
+	return summarizer
 }
 
-func (client *Client) constructSummary(vpas *v1beta2.VerticalPodAutoscalerList, excludeContainers string) (Summary, error) {
+func (s Summarizer) GetSummary() (Summary, error) {
 	var summary Summary
-	if len(vpas.Items) <= 0 {
-		return summary, nil
+	// cached vpas
+	if s.vpas == nil {
```

Comment:
It looks like the dashboard may be the only thing that makes use of the cache, but what happens if new VPAs get created after it creates the initial cache? Will the new ones get picked up? Should there be a TTL of the cache?

Reply:
Actually looking closer at the dashboard code I don't think anything will ever cache because a new Summarizer object gets created on every call. See: https://github.com/FairwindsOps/goldilocks/pull/154/files#diff-dc1496f734707825dbb6070e68866784R167

Reply:
This is what I think prompted me to change the caching and update behaviour of the Summarizer in the next set of commits where I go through and actually use the API in this package in the web ui and dashboard updates I make.